src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,960 @@
+  double o_range, o;
+  o_range = arma::datum::pi / n_orient;
+
+  CubeType HistArr(n_rbin, n_cbin, n_orient);
+  HistArr.zeros();
+
+  size_t r, c, o1, o2;
+  for(size_t i = 0; i < Magnitude.n_rows; ++i)
+  {
+    for(size_t j = 0; j < Magnitude.n_cols; ++j)
+    {
+      r = i / downscale;
+      c = j / downscale;
+
+      if( interp != 0)
+      {
+        o = Orientation(i, j) / o_range;
+        o1 = ((size_t) o) % n_orient;
+        o2 = (o1 + 1) % n_orient;
+        HistArr(r, c, o1) += Magnitude(i, j) * (1 + (int)o - o);
+        HistArr(r, c, o2) += Magnitude(i, j) * (o - (int) o);
+      }
+      else
+      {
+        o1 = (size_t) (Orientation(i, j) / o_range + 0.5) % n_orient;
+        HistArr(r, c, o1) += Magnitude(i, j);
+      }
+    }
+  }
+
+  HistArr = HistArr / downscale;
+
+  for (size_t i = 0; i < HistArr.n_slices; ++i)
+    HistArr.slice(i) = arma::square(HistArr.slice(i));
+  
+  return HistArr;
+}
+
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+GetShrunkChannels(CubeType const &InImage, CubeType &reg_ch, CubeType &ss_ch)
+{
+  CubeType luv = this->RGB2LUV(InImage);
+  
+  const size_t shrink = this->options["shrink"];
+  const size_t n_orient = this->options["n_orient"];
+  const size_t grd_smooth_rad = this->options["grd_smooth_rad"];
+  const size_t grd_norm_rad = this->options["grd_norm_rad"];
+  const size_t num_channels = 13;
+  const size_t rsize = luv.n_rows / shrink;
+  const size_t csize = luv.n_cols / shrink;
+  CubeType channels(rsize, csize, num_channels);
+  
+  
+  size_t slice_idx = 0;
+  
+  for( slice_idx = 0; slice_idx < luv.n_slices; ++slice_idx)
+    channels.slice(slice_idx)
+    = this->bilinearInterpolation(luv.slice(slice_idx), (size_t)rsize, (size_t)csize);
+
+  double scale = 0.5;
+  
+  while(scale <= 1.0)
+  {
+    CubeType img( (luv.n_rows * scale),
+                   (luv.n_cols * scale),
+                   luv.n_slices );
+
+    for( slice_idx = 0; slice_idx < luv.n_slices; ++slice_idx)
+    {
+      img.slice(slice_idx) = 
+        this->bilinearInterpolation(luv.slice(slice_idx), 
+                                    (luv.n_rows * scale),
+                                    (luv.n_cols * scale) );
+    }
+    
+    CubeType OutImage = this->ConvTriangle(img, grd_smooth_rad); 
+    
+    MatType Magnitude(InImage.n_rows, InImage.n_cols),
+             Orientation(InImage.n_rows, InImage.n_cols);
+    
+    this->Gradient(OutImage, Magnitude, Orientation);
+    
+    size_t downscale = std::max(1, (int)(shrink * scale));
+    
+    CubeType Hist = this->Histogram(Magnitude, Orientation,
+                                downscale, 0);
+
+    channels.slice(slice_idx) =
+    bilinearInterpolation( Magnitude, rsize, csize);
+    slice_idx++;
+    for(size_t i = 0; i < InImage.n_slices; ++i)
+      channels.slice(i + slice_idx) =
+      bilinearInterpolation( Magnitude, rsize, csize);    
+    slice_idx += 3;
+    scale += 0.5;
```

Comment:
I would recommend to use ``slice_idx += n_slices``instead of ``slice_idx += 3;`` there is no guarantee that we call the function with a 3 channel image.